server/svix-server/src/core/idempotency.rs

```diff
@@ -0,0 +1,426 @@
+// SPDX-FileCopyrightText: © 2022 Svix Authors
+// SPDX-License-Identifier: MIT
+
+//! Defines idempotency middleware for the Axum server which first looks up the given key for an
+//! existing response before routing to the given endpoint's function, and caches any such results
+//! such that subsequent requests to that endpoint with the same key will return the same response.
+//!
+//! Responses are cached for twelve hours by default.
+
+use std::{collections::HashMap, convert::Infallible, future::Future, pin::Pin, time::Duration};
+
+use axum::{
+    body::{Body, BoxBody, HttpBody},
+    http::{Request, Response, StatusCode},
+    response::IntoResponse,
+    Json,
+};
+use blake2::{Blake2b512, Digest};
+use serde::{Deserialize, Serialize};
+use tower::Service;
+
+use super::cache::{CacheKey, CacheValue, RedisCache};
+
+/// Returns the default exipry period for cached responses
+const fn expiry_default() -> Duration {
+    Duration::from_secs(60 * 60 * 12)
+}
+
+/// Returns the default expiry period for the starting lock
+const fn expiry_starting() -> Duration {
+    Duration::from_secs(20)
+}
+
+/// The data structure containing all necessary components of a response ready to be (de)serialized
+/// from/into the cache
+#[derive(Deserialize, Serialize)]
+enum SerializedResponse {
+    Start,
+    Finished {
+        code: u16,
+        headers: Option<HashMap<String, Vec<u8>>>,
+        body: Option<serde_json::Value>,
+    },
+}
+impl CacheValue for SerializedResponse {
+    type Key = IdempotencyKey;
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ConversionToResponseError {
+    #[error("the status code is out of bounds")]
+    StatusError(#[from] http::status::InvalidStatusCode),
+
+    #[error("a header name is invalid")]
+    FromStr(#[from] http::header::InvalidHeaderName),
+    #[error("a header value is invalid")]
+    InvalidHeaderValue(#[from] http::header::InvalidHeaderValue),
+}
+
+/// Will never error as long as Redis doesn't corrupt -- never use this with anything but values
+/// from Redis which were put in via the idempotency service from known good requests.
+fn finished_serialized_response_to_reponse(
+    code: u16,
+    headers: Option<HashMap<String, Vec<u8>>>,
+    body: Option<serde_json::Value>,
+) -> Result<Response<BoxBody>, ConversionToResponseError> {
+    let mut out = Json(body).into_response();
+
+    let status = out.status_mut();
+    *status = code.try_into()?;
+
+    if let Some(resp_headers) = headers {
+        let headers = out.headers_mut();
+        *headers = resp_headers
+            .iter()
+            .map(|(k, v)| Ok((k.parse()?, http::HeaderValue::from_bytes(v)?)))
+            .collect::<Result<_, ConversionToResponseError>>()?;
+    }
+
+    Ok(out)
+}
+
+struct IdempotencyKey(String);
+impl IdempotencyKey {
+    fn new(auth_token: &str, key: &str, url: &str) -> IdempotencyKey {
+        let mut hasher = Blake2b512::new();
+        hasher.update(format!("{}:{}:{}", auth_token, key, url));
```

Comment:
I think it's probably more efficient to do:
```suggestion
        hasher.update(auth_token);
        hasher.update(":");
        hasher.update(key);
        ...
```
So no string formatting.

Reply:
You haven't fixed it?